open_lm/attention.py

```diff
@@ -6,35 +6,65 @@
 
 
 def get_rectangular_mask(shape, q_seq_len, k_seq_len, device, dtype):
+    """
+    >>> get_rectangular_mask((1, 1), 2, 2, "cpu", torch.float32)
+    tensor([[[[0., -inf],
+              [0., 0.]]]])
+    >>> get_rectangular_mask((1, 1), 3, 5, "cpu", torch.float32)
+    tensor([[[[0., 0., 0., -inf, -inf],
+              [0., 0., 0., 0., -inf],
+              [0., 0., 0., 0., 0.]]]])
+    >>> get_rectangular_mask((1, 1), 5, 5, "cpu", torch.float32)
+    tensor([[[[0., -inf, -inf, -inf, -inf],
+              [0., 0., -inf, -inf, -inf],
+              [0., 0., 0., -inf, -inf],
+              [0., 0., 0., 0., -inf],
+              [0., 0., 0., 0., 0.]]]])
+    """
     # xformers requires the mask to be built with a shape that is a multiple of 8
     # probably because of the way it is implemented in CUDA
     next_multiple_8 = (k_seq_len + 7) // 8 * 8  #
-    mask = torch.ones((q_seq_len, next_multiple_8), device=device, dtype=bool)
+
+    mask = torch.ones((q_seq_len, k_seq_len), device=device, dtype=bool)
     mask[:, -q_seq_len:] = torch.tril(mask[:, -q_seq_len:], diagonal=0)
-    return torch.zeros((*shape, q_seq_len, next_multiple_8), device=device, dtype=dtype).masked_fill(
-        ~mask, float("-inf")
-    )[:, :, :, :k_seq_len]
+
+    output_mask = torch.zeros((*shape, q_seq_len, next_multiple_8), device=device, dtype=dtype)
+    output_mask[:, :, :, :k_seq_len].masked_fill_(~mask, torch.finfo(dtype).min)
+    return output_mask[:, :, :, :k_seq_len]
```

Comment:
I thought you needed to select -k_seq_len: here

Reply:
I fixed it above instead by setting up the first few columns of output_mask appropriately (instead of the last few columns)